src/spinlock.h

```diff
@@ -17,6 +19,7 @@ extern "C" {
 typedef struct spinlock_lock spinlock_lock_t;
 typedef _Volatile(spinlock_lock_t) spinlock_lock_volatile_t;
 struct spinlock_lock {
+    log_producer_t* log_producer_spinlock;
```

Comment:
Hey @fireuse the size of this struct can't be altered too much, it has to be 32 bits or smaller (4 bytes) otherwise the structs having to be cacheline-aligned will, literally, double up in size.

For example the struct hashtable_half_hashes_chunk can be only up to 64 bytes and the spinlock has up to only 4 bytes of space available otherwise to keep the alignment, that is necessary, a lot of empty padding will be added by the compiler basically doubling up the memory used.

For an hashtable with 2147483647 (2.147 billions) of elements this struct currently requires a bit less of 9.8GB of memory, increasing the padding would push the required memory almost 20GB and would basically slow down the execution by a time of 2 because the CPU would have to fetch always 2 cachelines.

For reference, the modern CPUs already fetches 2 cachelines, the requested one plus one in advance, and the platform relies on this to push the performances when searching into highly loaded hashtables.

Reply:
I think that having a static variable used internal by spinlock.c where the producer is defined would achieve the same result and greatly speed up the execution.

The hashtable doesn't initialise directly the spinlocks because the hashtable gets the memory allocated via MMAP and kernel zero the mmap-ed memory on page-faults by default, it means that the memory is instant-assigned to cachegrand and the kernel will take care of filling with zeroes the pages getting, as consequence, the spinlock initialised to zero as well.

Having to invoke spinlock_init internally would take some time quiet some time when working with big hashtable, for instance on an hashtable with a size of 2.147 billions of buckets there would be 153389363 chunks therefore it would have to be invoked that amount of times +32 because of the padding added at the end.

Reply:
The same approach may be used in the other places as well, the producer would be always the same inside the single file and it may change if we need performing an ad-hoc initialisation.

To use "constructors" in C you can rely on the constructor attribute, here more info
https://gcc.gnu.org/onlinedocs/gcc-9.3.0/gcc/Common-Function-Attributes.html#Common-Function-Attributes

Here an example
https://www.geeksforgeeks.org/__attribute__constructor-__attribute__destructor-syntaxes-c/#:~:text=__attribute__((constructor))%20runs,unloaded%2C%20typically%20at%20program%20exit.&text=__attribute__%20is%20a,a%20function%20or%20a%20macro.

Reply:
I didn't check but it's possible that altering that struct is also what is causing the coredump with the tests.